internal/storage/loopback/loopback.go

```diff
@@ -0,0 +1,191 @@
+		}
+
+		// Provision loop device
+		loopID = getLoopID(ctx, loopCtrl, device)
+		if loopID == -1 {
+			return errors.New("failed to provision loop device")
+		}
+	}
+
+	// Pair loop dev with backing file
+	source, err := pairLoopDevice(ctx, loopID, backingFile)
+	if err != nil {
+		return errors.Wrapf(err, "failed to pair backing file %s to loop device with ID %d", backingFile, loopID)
+	}
+
+	return unixMount(source, target, "ext4", uintptr(unix.MS_RDONLY), "noload")
+}
+
+// Teardown a file backed loop device
+func Teardown(ctx context.Context, device int) error {
+	if err := RemoveBackingFile(ctx, device); err != nil {
+		return errors.Wrap(err, "could not teardown loop device")
+	}
+
+	loopCtrl, err := getLoopController(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to get loop controller")
+	}
+
+	// remove loop device
+	deallocateLoopDevice(ctx, loopCtrl, device)
+	return nil
+}
+
+func getLoopController(ctx context.Context) (int, error) {
+	// Get loop controller
+	fd, err := safeOpen(ctx, "/dev/loop-control", unix.O_RDWR, EmptyMode)
+	if err != nil {
+		return 0, err
+	}
+	return fd, nil
+}
+
+func getLoopID(ctx context.Context, loopControllerFD, device int) int {
+	loopID, err := allocateLoopDevice(ctx, loopControllerFD, device)
+	if err != nil {
+		return -1
+	}
+	return loopID
+}
+
+// Implement some methods from https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=770fe30a46a12b6fb6b63fbe1737654d28e8484
+/*
+Note: devnr means device number
+Example in C:
+ cfd = open("/dev/loop-control", O_RDWR);
+
+ # add a new specific loop device
+ err = ioctl(cfd, LOOP_CTL_ADD, devnr);
+
+ # remove a specific loop device
+ err = ioctl(cfd, LOOP_CTL_REMOVE, devnr);
+
+ # find or allocate a free loop device to use
+ devnr = ioctl(cfd, LOOP_CTL_GET_FREE);
+
+ sprintf(loopname, "/dev/loop%i", devnr);
+ ffd = open("backing-file", O_RDWR);
+ lfd = open(loopname, O_RDWR);
+ err = ioctl(lfd, LOOP_SET_FD, ffd);
+*/
+
+func allocateLoopDevice(ctx context.Context, loopController, deviceNumber int) (int, error) {
+	device, err := safeIoctl(ctx, loopController, LOOP_CTL_ADD, deviceNumber)
+	if err == unix.EEXIST {
+		return deviceNumber, unix.EEXIST
+	}
+	return int(device), nil
+}
+
+func deallocateLoopDevice(ctx context.Context, loopController, deviceNumber int) int {
+	device, err := safeIoctl(ctx, loopController, LOOP_CTL_REMOVE, deviceNumber)
+	if err == unix.EBUSY {
+		log.G(ctx).WithField("deviceNumber", deviceNumber).Error("loop device is busy")
+		return int(device)
+	}
+
+	log.G(ctx).WithField("deviceNumber", device).Debug("Removed loop device")
+	return int(device)
+}
+
+// PairLoopDevice pairs a backing file `backingFile` to the loop device with device number
+// `deviceNumber`. Returns the loop device path, e.g. "/dev/loop9"
+func pairLoopDevice(ctx context.Context, deviceNumber int, backingFile string) (loopName string, err error) {
+	loopName = fmt.Sprintf("/dev/loop%d", deviceNumber)
```

Comment:
nit: consider passing in the loopName instead of the deviceNumber since we don't need it for anything else and the calling function already constructs the loopName.